### ISSUE TITLE
forcing file format compatible with TopoPyScale

### ISSUE THREAD

**Issue by ArcticSnow** [CryoGrid/CryoGrid.jl#9]
Hi @bgroenks96  and @jannitzbon,

As you may have heard at the last ITCH, we have been working on porting Toposcale to Python. This is now running well and publicly available: https://github.com/ArcticSnow/TopoPyScale

I had coded with Leo and Sebastian a routine to export the downscaled timeseries in a format compatible with the latest Cryogrid (Matlab). This is in the routine `to_cryogrid()`. Do you think we could put in place the same thing with your Julia implementation? For this we'd need to find out what is the best interface and format for passing over the data.

Subsequently, I hope using Cryogrid too, and why not the Julia version ;)  Julia is new to me though.

**Comment by bgroenks96** [comment]
Hi @ArcticSnow , that sounds great!

I haven't worked with the community model much so I am not aware of what file format it uses for forcings. Maybe you could educate me on this :)

For the Julia version, we have generally preferred standard file formats like CSV and JSON. This package currently provides a `TimeSeriesForcing` type that just takes a time series as input which can be loaded via the `JSON` format used by the "Lite" version of CryoGrid.

I would like to also have a forcing type that loads from spatially distributed formats like NetCDF or similar, so what you propose sounds like it would also accomplish this.

More parity in input formats with the community model would also be good and has always been the goal! Just not a priority :(

**Comment by ArcticSnow** [comment]
Ok great! 
So we do not need to match the format of the community model but establish and find the optimal way to exchange data from the downscaling toolbox and cryogrid.jl.

The easiest from `topopyscale` is to export in netcdf format (as we can also compress). All I need is the list of variables needed and their respective unit. Not sure how much time you'd have, but if writing an option supporting netcdf in `loadforcing()` function is easy to you, I think we can have it working shortly. 

With `TopoPyScale` we abstract a region using clustering based on morphometric factors (something we call `toposub`. This groups pixels of a DEM with similar attributes and therefore reduces the number of simulation to perform significantly. So by default, we extract one file per cluster centroids containing all the timeseries. I can send you by email a sample of what for instance we use with the community model. But once again if you think there is better way, everything is possible.

**Comment by bgroenks96** [comment]
I would also consider refactoring the existing code a bit since `loadforcings` currently doesn't return the actual `Forcing` type, which is actually kind of confusing.... this would be a good opportunity to improve this. Ideally I think it should initialize the appropriate `Forcing` implementation based on the file type (e.g. `TimeSeriesForcing` for JSON or CSV, and some other `struct` for a NetCDF forcing from TopoPyScale).

What does the `to_cryogrid` function actually do for the community model? Is the output also NetCDF? Or is it some special format?

As for units, I suggest that we standardize this between all CryoGrid versions for simplicity.

**Comment by jannitzbon** [comment]
Hi @ArcticSnow ,
I don't have something to add regarding CryoGrid.jl, but I am currently implementing a fork of `CryoGridLite.jl` within AWI's ESM. For this, I am also using `netCDF` as file format.
One thing which increasingly annoyed me about  the old CryoGrid3 forcing data (I guess it is the same though for the community model), is that the time information was stored as MATLAB datenum, which I think is even within MATLAB already depreciated. So it would be great to have the time information as a datetime object which would improve the interoperatbility with the julia version.

**Comment by bgroenks96** [comment]
I agree @jannitzbon, that's why I have two `JSON` formats supported in CryoGrid.jl. The first is Moritz's version that uses DateNum, and the second uses a standard time format (UNIX time, I think?).

**Comment by ArcticSnow** [comment]
so `to_cryogrid()` function creates a netcdf file with time as a coordinate. It is created with xarray so that it is a datetime object. Seb had written a function to read this netcdf within the community model. 

https://github.com/ArcticSnow/TopoPyScale/blob/0bb2e2489600913306fb749a898f413316801bff/TopoPyScale/topo_export.py#L140

@jannitzbon do you want me to send you sample files? Bryan got them already

**Comment by bgroenks96** [comment]
Hi @ArcticSnow, 

This is currently in progress on a local branch of mine. I am implementing a generic NetCDF loader in addition to the TopoPyScale specific one together.

I really like that TopoPyScale includes the unit information; this is super helpful. I noticed, however, that it uses a somewhat strange notation with powers (e.g. "W/m2") that doesn't work with `Unitful.jl` which CryoGrid.jl uses for unit types. Would it be a terribly inconvenient change to TopoPyScale to change this convention to `W/m^2` or similar?

**Comment by ArcticSnow** [comment]
Hi Brian,
I just pushed a commit [b0060f2](https://github.com/ArcticSnow/TopoPyScale/commit/b0060f2d52b0e63784587cd35a913b564a215b4e). Let me know if that is enough or more is needed for compatibility?

**Comment by bgroenks96** [comment]
@ArcticSnow It would also be good for `C` to be `°C` and `mm/day` to be `mm/d`, but it's not strictly  necessary as I have handled these cases manually in the current loader implementation.

**Comment by ArcticSnow** [comment]
ok, done. :)

**Comment by bgroenks96** [comment]
Could you send me some updated forcing files?

**Comment by ArcticSnow** [comment]
sent via email.

**Comment by bgroenks96** [comment]
@ArcticSnow What is the purpose of the `add_offset` and `scale_factor` attributes? Do these need to be applied to the forcing data before use? I assume not because `rainfall` and `snowfall` already have minimum values of zero.

**Comment by bgroenks96** [comment]
Hmm, well then I guess `NCDatasets.jl` must handle it correctly. `NetCDF.jl` did not, though.

**Comment by ArcticSnow** [comment]
I know that the Matlab Netcdf reader handles it too as Leo and Seb had no prb. In Julia, I have no clue

**Comment by bgroenks96** [comment]
Ok, something is wrong. The values loaded by `NCDatasets` and `xarray` don't match...

**Comment by ArcticSnow** [comment]
I installed `NCDatasets.jl` and I seem to have same output in both `xarray` from Python or `NCDatasets.jl` from Julia 
![image](https://user-images.githubusercontent.com/2042458/236685695-87959471-f221-4549-9420-72fd2a1793b0.png)

**Comment by bgroenks96** [comment]
Are they exactly the same? Or just visually?

**Comment by bgroenks96** [comment]
Wait nevermind, it's OK, I just forgot to use 0-based indexing for python.... :facepalm:

**Comment by ArcticSnow** [comment]
![image](https://user-images.githubusercontent.com/2042458/236686030-296cde89-8650-46fd-ad8c-ef4f12dc082a.png)

**Comment by bgroenks96** [comment]
Does the community model currently load the full forcing file into memory? Or does it read lazily from disk?

I  think the latter would be preferable for large scale simulations but it's somewhat non-trivial to implement.

**Comment by ArcticSnow** [comment]
![image](https://user-images.githubusercontent.com/2042458/236686084-71c2762d-6c4d-4b55-b149-c5cb01f9853b.png)

**Comment by ArcticSnow** [comment]
I have no clue about how Matlab handles this. Leo would know better. I'll ask him.
In Python, xarray loads by default lazily.

**Comment by ArcticSnow** [comment]
But the current workflow I have heard off is to run one forcing file at a time, so one timeseries. Then he launch in parallel many more independently. and later combine all run together using the cluster map (in case of clustering the DEM for spatialized modeling).

**Comment by bgroenks96** [comment]
I did a quick survey of the current state of lazy disk array processing in Julia and it's sadly still very much in its infancy. There is a package called `DiskArrays` which is under development and is used by `NetCDF.jl` but `NCDatasets.jl` doesn't support it yet (see [this issue](https://github.com/Alexander-Barth/NCDatasets.jl/issues/79)).

So I guess we'll settle for the memory hungry solution for now. If it becomes a problem, I can look into this further.

**Comment by ArcticSnow** [comment]
Thank you @bgroenks96 for the effort to make `Cryogrid.jl` compatible with `TopoPyScale` output files!! I am looking forward to trying this new modeling pipeline for my current study. As I have to edit the documentation, I'll add a pointer to the Julia implementation of Cryogrid. We could even think eventually of a full example combining both tools in a Binder notebook.

**Comment by bgroenks96** [comment]
@ArcticSnow Cool, sounds good. What is your target use case for CryoGrid? Bear in mind that CryoGrid.jl has only a subset of the functionality insofar as physical processes; e.g. there is only a very simple snow scheme at the moment and no vegetation physics at all. There is also no lateral coupling though I hope to add this in the near future.

1D heat conduction (in several forms), water transport (saturated and unsaturated flow), and the surface energy balance are all available. The fast CryoGridLite implicit solver scheme is also implemented.
